### DISCUSSION TITLE
Template conditional logic

### DISCUSSION THREAD

**Discussion by simsrw73** [JanDeDobbeleer/oh-my-posh#3249] [Troubleshoot] [answered]
I'm trying to create a new theme based on Catppuccin color theme: [https://github.com/simsrw73/poshcat.omp](https://github.com/simsrw73/poshcat.omp)

I want to have a segment that displays the Project Name. I'm thinking of having it derive the project name from the Project segment if the necessary files are present. If not fall back to deriving it from git. And finally, falling back to displaying nothing if neither of those work. But I can't get the IF logic to work as expected. I've tried several permutations, but it never works. I can access Segments.Git.RepoName but it never reaches a branch where it displays. I'm doing something stupid, but I've been staring at it so long I can't see it. How do I do this?

```
{
    "background": "p:Flamingo",
    "foreground": "p:Crust",
    "powerline_symbol": "\ue0b4",
    "style": "powerline",
    "template": " {{ if .Error }}Why can't I get here?{{ else if .Name }}{{ .Name }}{{ else }}Or here?{{ end }}",
    "type": "project"
},
```

**Comment by JanDeDobbeleer** [comment, marked as answer]
Hey @simsrw73, let me see if I can help you out. First, looking at the [project properties](https://ohmyposh.dev/docs/segments/project#properties), there's no `.Error` property. You could use the following template:

```json
"template": " {{ if .Name }}{{ .Name }}{{ else }}No name{{ end }} "
```

**Reply by simsrw73** [reply]
Thanks. I've played around a bit more and got "something" working. Three possible bugs.

- I copied that .Error check from the docs. In both of the code blocks on the Project Segment page.
- If the Project segment fails to find any files, it never evaluates the Template argument. It omits displaying the segment, so the else clause is never evaled. I ended up using an additional segment, a Text seg, to display the derived project name. ([https://github.com/simsrw73/poshcat.omp](https://github.com/simsrw73/poshcat.omp))
- In the Project segment, to get it to display nothing, I had to set `"template": " "`, a single space. Setting it to empty `""` caused it to display a default template.

**Reply by simsrw73** [reply]
An additional observation. The docs say that for cross-segment properties to be used, the segment must appear in the config. At least in this case, it is untrue. If I delete the Project segment entirely, I can still access `.Segments.Project.Name`

I could argue this one either way. It would be nice if mentioning a Segment property caused it to become available. Lazy evaluation. No files are searched and read for `.Segments.Project.Name` unless it is referenced. On the other hand, these ops are relatively expensive, so it makes sense to require the user to explicitly declare they know the cost and want the functionality. Either way, my observation and the docs do not agree: I can access `.Segments.Project.Name` without having a Project Segment in contradiction to the docs.

**Reply by JanDeDobbeleer** [reply]
@simsrw73 only because that's still cached, there's no logic in place that will fetch it when the segment isn't explicitly there.

**Reply by JanDeDobbeleer** [reply]
> * I copied that .Error check from the docs. In both of the code blocks on the Project Segment page.

Ah, yes. It wasn't added to the docs, fixed that.

> * If the Project segment fails to find any files, it never evaluates the Template argument. It omits displaying the segment, so the else clause is never evaluated. I ended up using an additional segment, a Text segment, to display the derived project name. (https://github.com/simsrw73/poshcat.omp)

I can add `always_enabled` just like we do for others to avoid the need for multiple segments.

> * In the Project segment, to get it to display nothing, I had to set `"template": " "`, a single space. Setting it to empty `""` caused it to display a default template.

Correct, an empty `template` will always fallback to the default one.
